### PR TITLE
Support DeleteResourceOnFinalize k8s plugin config

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

# TL;DR
Implement DeleteResourceOnFinalize to delete k8s resources on Finalize() this ensures resources do not continue to consume cluster resources after flyte has detected they completed.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1152

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -32,4 +32,7 @@ require (
 	sigs.k8s.io/controller-runtime v0.8.2
 )
 
-replace github.com/aws/amazon-sagemaker-operator-for-k8s => github.com/aws/amazon-sagemaker-operator-for-k8s v1.0.1-0.20210303003444-0fb33b1fd49d
+replace (
+	github.com/aws/amazon-sagemaker-operator-for-k8s => github.com/aws/amazon-sagemaker-operator-for-k8s v1.0.1-0.20210303003444-0fb33b1fd49d
+	github.com/flyteorg/flyteplugins => github.com/flyteorg/flyteplugins v0.5.55-0.20210616003554-515cc6e2fded
```
Comment (EngHabu) [inline go.mod:37]: Reminder to revert after the other PR is merged

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 2]
```diff
@@ -361,18 +361,25 @@ func (e *PluginManager) ClearFinalizers(ctx context.Context, o client.Object) er
...
 			logger.Errorf(ctx, "Failed to build the Resource with name: %v. Error: %v, when finalizing.", tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), err)
 			return nil
 		}
 
 		e.AddObjectMetadata(tCtx.TaskExecutionMetadata(), o, config.GetK8sPluginConfig())
-		nsName := k8stypes.NamespacedName{Namespace: o.GetNamespace(), Name: o.GetName()}
+		nsName = k8stypes.NamespacedName{Namespace: o.GetNamespace(), Name: o.GetName()}
+	}
+
+	// If you change InjectFinalizer on the
```
Comment (katrogan) [inline pkg/controller/nodes/task/k8s/plugin_manager.go:381]: ?

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 3]
```diff
@@ -389,10 +396,26 @@ func (e *PluginManager) Finalize(ctx context.Context, tCtx pluginsCore.TaskExecu
...
+	// If we should delete the resource when finalize is called, do a best effort delete.
+	if config.GetK8sPluginConfig().DeleteResourceOnFinalize {
+		// Attempt to delete resource, if not found, return success.
+		if err := e.kubeClient.GetClient().Delete(ctx, o); err != nil {
+			if IsK8sObjectNotExists(err) {
+				return errs.ErrorOrDefault()
+			}
+
+			// This happens sometimes because a node gets removed and K8s deletes the pod. This will result in a
+			// Pod does not exist error. This should be retried using the retry policy
```
Comment (katrogan) [inline pkg/controller/nodes/task/k8s/plugin_manager.go:412]: > This should be retried using the retry policy
is this descriptive or meant to be a todo?

**Review by katrogan — APPROVED** [review]
thank you for adding this!

File: pkg/controller/nodes/task/k8s/plugin_manager.go (hunk 3 above)
Comment (kumare3) [reply pkg/controller/nodes/task/k8s/plugin_manager.go:412]: Should we make this a plugin property or just blanket?

Comment (EngHabu) [reply pkg/controller/nodes/task/k8s/plugin_manager.go:412]: I guess we can offer an override per plugin? I was debating doing that or whether that was an overkill.. but yeah we can add that...

Comment (EngHabu) [reply pkg/controller/nodes/task/k8s/plugin_manager.go:412]: Added [here](https://github.com/flyteorg/flyteplugins/pull/184/files)

File: pkg/controller/nodes/task/k8s/plugin_manager.go (hunk 2 above)
Comment (EngHabu) [reply pkg/controller/nodes/task/k8s/plugin_manager.go:381]: In my defense, it was like that, I just moved code around :D.... trying to infer intent

Comment (katrogan) [reply pkg/controller/nodes/task/k8s/plugin_manager.go:381]: oh, fair. but probably nice to clean up :)

**Review by katrogan — APPROVED** [review]

**Review by katrogan — APPROVED** [review]

**Comment by katrogan** [comment]
don't you still need to update go.mod?

**Review by katrogan — APPROVED** [review]

**Comment by EngHabu** [comment]
oh thank you. I ran `go get -u githbu.com/flyteorg/flyteplugins` but that doesn't seem to have worked... I should have verified. thanks for catching it

**Pull request merged**